internal/libs/strings/string.go

```diff
@@ -3,8 +3,64 @@ package strings
 import (
 	"fmt"
 	"strings"
+
+	tmbytes "github.com/tendermint/tendermint/libs/bytes"
 )
 
+type lazyStringf struct {
+	tmpl string
+	args []interface{}
+	out  string
+}
+
+func (s *lazyStringf) String() string {
+	if s.out == "" && s.tmpl != "" {
+		s.out = fmt.Sprintf(s.tmpl, s.args)
+		s.args = nil
+		s.tmpl = ""
+	}
+	return s.out
+}
+
+func LazySprintf(t string, args ...interface{}) fmt.Stringer {
```

Comment:
I think we need some documentation for this, including a big warning reminding people that the result captures the arguments and is not thread-safe.